oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,344 @@
+  (``doc.in``)
+* Additional utilities needed to perform common development tasks (``dev.in``)
+* Utilities that a particular developer likes to use with a repository, but
+  aren't strictly needed for any of the regular contexts (``private.in``).
+
+.. _Sphinx: http://www.sphinx-doc.org/
+
+Declare Direct Dependencies
+---------------------------
+
+As indicated above, some of the usage contexts have a standard filename used in
+the ``requirements`` directory of an Open edX repository to list dependencies.
+Others will have an appropriate filename custom to that repository's unique
+context.  Each of these is a ``pip``-compatible `requirements file`_ listing
+the direct dependencies needed for that context.  Beyond complying with the
+file format, there are a few guidelines each of these files should follow:
+
+* The file should start with a brief comment explaining the context in which
+  these dependencies are needed.  Examples can be found in the
+  `cookiecutter-django-app`_ repository.
+* Each listed dependency should have a brief end-of-line comment explaining
+  its primary purpose(s) in this context.  These comments typically start at
+  the 27th character, but this is just a convention for consistency with files
+  generated by ``pip-compile``.
+* Version constraints should only be used to exclude dependency versions which
+  are known (or strongly suspected) to not work in this context.
+* Indirect dependencies (used by dependencies but not directly by the code in
+  the repository itself) should not be listed unless a constraint is needed to
+  enforce a compatible version; these are automatically detected and captured
+  elsewhere as described below.
+* `Environment markers`_ should be used as necessary to indicate dependencies
+  which should only be installed on specific operating systems, Python
+  versions, etc.
+* If the dependencies in one context are a superset of those in another one,
+  do not repeat the dependencies.  Instead, explicitly include the file
+  listing the common dependencies in the superset context file (e.g.
+  ``-r base.txt``) and explain in an end-of-line comment why that set of
+  dependencies is also needed in this context.  Note that the ``pip-compile``
+  output file should be included rather than the looser requirements file it
+  was generated from, as this ensures that the same versions of packages are
+  installed in the different contexts.  We don't want to run tests with
+  different versions of dependencies than we use in production, for example.
+* Avoid direct links to packages local directories, GitHub, or other version
+  control systems if at all possible; all dependencies should be installed
+  from `PyPI`_.  If you think you're in one of the rare circumstances where
+  installing a package from a URL is appropriate, see the notes below on
+  `Installing Dependencies from URLs`_
+
+If the repository contains a Python package, base dependencies also need to
```

Comment:
What else could a repository contain?  IDAs don't have setup.py I guess?  Maybe better is:  "if the repository contains a setup.py"

Reply:
Sure, how about: "If the repository contains a `setup.py` file defining a Python package, the
base dependencies also need to be specified there."